enhancements/improve-kubeletconfig-default-configuration-check-enhancement.md

```diff
@@ -0,0 +1,278 @@
+		Filter:   `.kubeletconfig|.kind="KubeletConfiguration"|.apiVersion="kubelet.config.k8s.io/v1beta1"`,
+	})
+}
+```
+
+example result
+```json
+{
+  "tlsCipherSuites": [
+    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
+    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
+    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
+    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
+    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
+  ],
+  "tlsMinVersion": "VersionTLS12",
+  "rotateCertificates": true,
+  "serverTLSBootstrap": true,
+  "registryPullQPS": 5,
+  "registryBurst": 10,
+  "eventRecordQPS": 5,
+  "eventBurst": 10,
+  "enableDebuggingHandlers": true,
+  "streamingConnectionIdleTimeout": "4h0m0s",
+  "cpuManagerReconcilePeriod": "10s",
+  "memoryManagerPolicy": "None",
+  "topologyManagerPolicy": "none",
+  "topologyManagerScope": "container",
+  "runtimeRequestTimeout": "2m0s",
+  "hairpinMode": "promiscuous-bridge",
+  "maxPods": 250,
+  "podPidsLimit": 4096,
+  "resolvConf": "/etc/resolv.conf",
+  "cpuCFSQuota": true,
+  "cpuCFSQuotaPeriod": "100ms",
+  "nodeStatusMaxImages": 50,
+  "maxOpenFiles": 1000000,
+  "contentType": "application/vnd.kubernetes.protobuf",
+  "kubeAPIQPS": 50,
+  "kubeAPIBurst": 100,
+  "serializeImagePulls": false,
+  "evictionHard": {
+    "imagefs.available": "15%",
+    "memory.available": "100Mi",
+    "nodefs.available": "10%",
+    "nodefs.inodesFree": "5%"
+  },
+  "evictionPressureTransitionPeriod": "5m0s",
+  "enableControllerAttachDetach": true,
+  "makeIPTablesUtilChains": true,
+  "iptablesMasqueradeBit": 14,
+  "iptablesDropBit": 15,
+  "featureGates": {
+    "APIPriorityAndFairness": true,
+    "CSIMigrationAWS": false,
+    "CSIMigrationAzureFile": false,
+    "CSIMigrationGCE": false,
+    "CSIMigrationvSphere": false,
+    "DownwardAPIHugePages": true,
+    "PodSecurity": true,
+    "RotateKubeletServerCertificate": true
+  },
+  "failSwapOn": true,
+  "memorySwap": {},
+  "containerLogMaxSize": "50Mi",
+  "containerLogMaxFiles": 5,
+  "configMapAndSecretChangeDetectionStrategy": "Watch",
+  "systemReserved": {
+    "cpu": "500m",
+    "memory": "1Gi"
+  },
+  "enforceNodeAllocatable": [
+    "pods"
+  ],
+  "volumePluginDir": "/etc/kubernetes/kubelet-plugins/volume/exec",
+  "logging": {
+    "format": "text",
+    "flushFrequency": 5000000000,
+    "verbosity": 2,
+    "options": {
+      "json": {
+        "infoBufferSize": "0"
+      }
+    }
+  },
+  "enableSystemLogHandler": true,
+  "shutdownGracePeriod": "0s",
+  "shutdownGracePeriodCriticalPods": "0s",
+  "enableProfilingHandler": true,
+  "enableDebugFlagsHandler": true,
+  "seccompDefault": false,
+  "registerNode": true,
+  "kind": "KubeletConfiguration",
+  "apiVersion": "kubelet.config.k8s.io/v1beta1"
+}
+```
+
+When we get `KubeletConfig` for each node, we will compare them, and
+only save the consistent result to a file at `/api/v1/nodes/proxy/configz`.
```

Comment:
Two questions:
1. is it going to be one result per cluster or one per pool? Can the configs differ between pools? What about overlapping pools or sub-pools?
2. What if there is a difference between the configs?

Reply:
1. I think we should save the kubeletconfig per pool, and then save it to`/api/v1/nodes/proxy/configz/{{pool}}`, and in the rule, we can ask the user to set a variable to choose which pool to run the scan on.
2. We will only keep parts of configuration which are the same across the pool.

Reply:
> 1. I think we should save the kubeletconfig per pool, and then save it to`/api/v1/nodes/proxy/configz/{{pool}}`, and in the rule, we can ask the user to set a variable to choose which pool to run the scan on.

Yes, this is what I wanted us to get to. I think per-pool is fine, one global kubelet config would not.

> 
>     2. We will only keep parts of configuration which are the same across the pool.

I would suggest to fail, to be honest, if the differences are something we care about.

Reply:
The rule will fail because the configuration needed for the check can't be found in the config file

Reply:
As long as we fail, I'm fine. Thank you.